src/instance.rs

```diff
@@ -837,20 +850,17 @@ impl<'py, T> Bound<'py, T> {
     /// Casts this `Bound<T>` to a `Borrowed<T>` smart pointer.
     #[inline]
     pub fn as_borrowed<'a>(&'a self) -> Borrowed<'a, 'py, T> {
-        Borrowed(
-            unsafe { NonNull::new_unchecked(self.as_ptr()) },
-            PhantomData,
-            self.py(),
-        )
+        // SAFETY: self is known to be a valid pointer to T and will be borrowed from the lifetime 'a
+        unsafe { Borrowed::from_non_null(self.py(), (self.1).0).cast_unchecked() }
     }
 
     /// Removes the connection for this `Bound<T>` from the GIL, allowing
     /// it to cross thread boundaries.
     #[inline]
     pub fn unbind(self) -> Py<T> {
-        // Safety: the type T is known to be correct and the ownership of the
-        // pointer is transferred to the new Py<T> instance.
         let non_null = (ManuallyDrop::new(self).1).0;
+        // SAFETY: the type T is known to be correct and the `ManuallyDrop` ensures
+        // the ownership of the reference is transferred into the `Py<T>`.
         unsafe { Py::from_non_null(non_null) }
```

Comment:
I think we can use `ManuallyDrop::into_inner` here which would make this safe.

Reply:
I was thinking we could maybe just take out the `ManuallyDrop<Py<T>>` stored within the `Bound` directly.

Reply:
Ah, right. We are not allowed to move out of it, because it implements `Drop` (I always forget about that...). I think this fine then.